src/tbp/plot/plots/interactive_hypothesis_space_pointcloud.py

```diff
@@ -0,0 +1,1717 @@
+                DataLocatorStep.index(name="step"),
+            ]
+        )
+
+        return locators
+
+    def remove(self, widget: Mesh) -> None:
+        if widget is not None:
+            self.plotter.at(1).remove(widget)
+            self.plotter.at(1).render()
+
+    def update_mesh(
+        self, widget: Mesh | None, msgs: list[TopicMessage]
+    ) -> tuple[Mesh | None, bool]:
+        """Update the target mesh when the episode changes.
+
+        Removes any existing mesh, loads the episode's primary target object,
+        applies its Euler rotations, scales and positions it, then adds it to
+        the plotter.
+
+        Args:
+            widget: The currently displayed mesh, if any.
+            msgs: Messages received from the `WidgetUpdater`.
+
+        Returns:
+            A tuple `(mesh, False)`. The second value is `False` to indicate
+            that no publish should occur.
+        """
+        self.remove(widget)
+        msgs_dict = {msg.name: msg.value for msg in msgs}
+
+        locator = self._locators["target"]
+        target = self.data_parser.extract(
+            locator, episode=str(msgs_dict["episode_number"])
+        )
+        target_id = target["primary_target_object"]
+        target_rot = target["primary_target_rotation_euler"]
+        target_pos = target["primary_target_position"]
+        widget = self.ycb_loader.create_mesh(target_id).clone(deep=True)
+        widget.rotate_x(target_rot[0])
+        widget.rotate_y(target_rot[1])
+        widget.rotate_z(target_rot[2])
+        widget.shift(*target_pos)
+        widget.alpha(1.0 - msgs_dict["transparency_value"])
+
+        self.plotter.at(1).add(widget)
+        self.plotter.at(1).render()
+
+        return widget, False
+
+    def update_agent(
+        self, widget: Mesh | None, msgs: list[TopicMessage]
+    ) -> tuple[Mesh | None, bool]:
+        """Update the agent and sensor patch location on the object.
+
+        Args:
+            widget: The currently displayed mesh, if any.
+            msgs: Messages received from the `WidgetUpdater`.
+
+        Returns:
+            A tuple `(widget, False)`. The second value is `False` to indicate
+            that no publish should occur.
+        """
+        msgs_dict = {msg.name: msg.value for msg in msgs}
+        episode_number = msgs_dict["episode_number"]
+        step_number = msgs_dict["step_number"]
+
+        steps_mask = self.data_parser.extract(
+            self._locators["steps_mask"], episode=str(episode_number)
+        )
+        mapping = np.flatnonzero(steps_mask)
+
+        agent_pos = self.data_parser.extract(
+            self._locators["agent_location"],
+            episode=str(episode_number),
+            sm_step=int(mapping[step_number]),
+        )
+
+        patch_pos = self.data_parser.extract(
+            self._locators["patch_location"],
+            episode=str(episode_number),
+            step=step_number,
```

Comment:
Is there a reason that the agent-pose uses `int(mapping[step_number])`, while here we just use `step_number`?

Reply:
Makes sense, thanks!